hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenRejectHandler.java

```diff
@@ -0,0 +1,380 @@
+        validateTrue(tokenRelation.balance() > 0, INSUFFICIENT_TOKEN_BALANCE);
+
+        return createFungibleTransfer(tokenId, accountID, tokenRelation.balance(), token.treasuryAccountId());
+    }
+
+    /**
+     * Updates the NFT allowances in the state after the token transfers have been processed,
+     * by setting the spender to null.
+     *
+     * @param tokenReferences The list of tokens that were rejected.
+     * @param nftStore The store to access writable NFT information.
+     */
+    private void updateNFTAllowances(
+            @NonNull final List<TokenReference> tokenReferences, @NonNull final WritableNftStore nftStore) {
+        tokenReferences.stream()
+                .filter(TokenReference::hasNft)
+                .map(reference -> getIfUsable(reference.nftOrThrow(), nftStore))
+                .forEach(nft -> {
+                    if (nft.hasSpenderId()) {
+                        nftStore.put(
+                                nft.copyBuilder().spenderId((AccountID) null).build());
+                    }
+                });
+    }
+
+    /**
+     * Updates the fungible token allowances in the state after the token transfers to the treasuries have been processed.
+     *
+     * @param rejectingAccount The account rejecting the tokens.
+     * @param tokenReferences The list of tokens that were rejected.
+     * @param accountStore The store to access writable account information.
+     */
+    private void updateFungibleAllowances(
+            @NonNull final Account rejectingAccount,
+            @NonNull final List<TokenReference> tokenReferences,
+            @NonNull final WritableAccountStore accountStore) {
+        final var updatedFungibleTokenAllowances = new ArrayList<>(rejectingAccount.tokenAllowances());
+        updatedFungibleTokenAllowances.removeIf(allowance -> tokenReferences.stream()
+                .filter(TokenReference::hasFungibleToken)
+                .anyMatch(fungibleToken -> Objects.equals(fungibleToken.fungibleToken(), allowance.tokenId())));
+
+        final var updatedAccount = rejectingAccount
+                .copyBuilder()
+                .tokenAllowances(updatedFungibleTokenAllowances)
+                .build();
+        accountStore.put(updatedAccount);
+    }
+
+    @NonNull
+    @Override
+    public Fees calculateFees(final FeeContext feeContext) {
+        final var body = feeContext.body();
+        final var op = body.tokenRejectOrThrow();
+        final var config = feeContext.configuration();
+        final var tokenMultiplier =
+                requireNonNull(config).getConfigData(FeesConfig.class).tokenTransferUsageMultiplier();
+
+        final int totalRejections = op.rejections().size();
+        int numOfFungibleTokenRejections = 0;
+        int numOfNFTRejections = 0;
+        for (final var rejection : op.rejections()) {
+            if (rejection.hasFungibleToken()) {
+                // Each fungible token rejection involves 2 AccountAmount transfers
```

Comment:
Do we have a check anywhere to make sure each token. rejection has 2 AccountAmount transfers ?

Reply:
Hey! 
Our only use case is:
Bob rejects token A he owns and the treasury account Carol receives the whole token A balance of Bob.

And since:
1 TokenTransfer = 1 TokenReference.
1 Token Transfer body contains List<AccountAmounts>

We always have exactly 2 AccountAmounts, because we have 2 different accounts that exchange tokens.

Reply:
@MrValioBg Since user can submit any invalid transactions, do we reject any transaction that is not like the valid use case ?

Reply:
Users can only specify the token they want to reject as a TokenReference. We have checks if they actually own it. If they don't have a balance of it, the TokenReject fails.

Reply:
( They don't specify treasury accounts, we find which is the treasury account)